### ISSUE TITLE
systemd-resolved[23]: Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:691, function hashmap_iterate_in_internal_order(). Aborting.

### ISSUE THREAD

**Issue by Keruspe** [systemd/systemd#943]
I was doing some tests with nspawn and while investigating another issue I saw this in the journal of a container.

systemd 224

**Comment by richardmaw-codethink** [comment]
Just to double check, are you running v224 inside the container, or outside of it, or both?

**Comment by Keruspe** [comment]
Actually, 224 inside the container and systemd git from yesterday on the host

**Comment by poettering** [comment]
Any chance you can get us a backtrace for this?

Most likely this broke with one of @teg's most recent changes.

**Comment by Keruspe** [comment]
Unfortunately, no coredump listed on the container nor the host by coredumpctl

**Comment by teg** [comment]
Is it reproducible? Any chance you could bisect?

**Comment by Keruspe** [comment]
Tried to reproduce a couple of times without success :/
Probably some kind of rare race?

**Comment by Keruspe** [comment]
Oh, re reading the logs, the logs just before might be helpful actually.
Here is the whole log extract from resolved from the boot when it failed

```
-- Reboot --
Aug 13 14:54:50 localhost systemd[1]: Starting Network Name Resolution...
Aug 13 14:54:50 localhost systemd-resolved[23]: Using system hostname 'localhost'.
Aug 13 14:54:50 localhost systemd-resolved[23]: There appears to be another LLMNR responder running. Turning off LLMNR support.
Aug 13 14:54:50 localhost systemd-resolved[23]: Switching to system DNS server 192.168.42.254.
Aug 13 14:54:50 localhost systemd[1]: Started Network Name Resolution.
Aug 13 14:54:52 localhost systemd-resolved[23]: Detected conflict on localhost IN AAAA  2a01:e35:2e93:2040:2540:5d0d:f2fa:efe9
Aug 13 14:54:52 localhost systemd-resolved[23]: Hostname conflict, changing published hostname from 'localhost' to 'localhost4'.
Aug 13 14:54:52 localhost systemd-resolved[23]: Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:691, function hashmap_iterate_in_internal_order(). Aborting.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Main process exited, code=killed, status=6/ABRT
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Unit entered failed state.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Failed with result 'signal'.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Service has no hold-off time, scheduling restart.
Aug 13 14:54:52 localhost systemd[1]: Stopped Network Name Resolution.
Aug 13 14:54:52 localhost systemd[1]: Starting Network Name Resolution...
Aug 13 14:54:52 localhost systemd-resolved[31]: Using system hostname 'localhost'.
Aug 13 14:54:52 localhost systemd-resolved[31]: There appears to be another LLMNR responder running. Turning off LLMNR support.
Aug 13 14:54:52 localhost systemd-resolved[31]: Switching to system DNS server 192.168.42.254.
Aug 13 14:54:52 localhost systemd[1]: Started Network Name Resolution.
Aug 13 14:54:52 localhost systemd-resolved[31]: Detected conflict on localhost IN AAAA  2a01:e35:2e93:2040:2540:5d0d:f2fa:efe9
Aug 13 14:54:52 localhost systemd-resolved[31]: Hostname conflict, changing published hostname from 'localhost' to 'localhost10'.
Aug 13 14:54:52 localhost systemd-resolved[31]: Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:691, function hashmap_iterate_in_internal_order(). Aborting.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Main process exited, code=killed, status=6/ABRT
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Unit entered failed state.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Failed with result 'signal'.
Aug 13 14:54:52 localhost systemd[1]: systemd-resolved.service: Service has no hold-off time, scheduling restart.
Aug 13 14:54:52 localhost systemd[1]: Stopped Network Name Resolution.
Aug 13 14:54:52 localhost systemd[1]: Starting Network Name Resolution...
Aug 13 14:54:52 localhost systemd-resolved[32]: Using system hostname 'localhost'.
Aug 13 14:54:52 localhost systemd-resolved[32]: There appears to be another LLMNR responder running. Turning off LLMNR support.
Aug 13 14:54:52 localhost systemd-resolved[32]: Switching to system DNS server 192.168.42.254.
Aug 13 14:54:52 localhost systemd[1]: Started Network Name Resolution.
Aug 13 14:54:52 localhost systemd-resolved[32]: Detected conflict on localhost IN A     192.168.42.39
Aug 13 14:54:52 localhost systemd-resolved[32]: Hostname conflict, changing published hostname from 'localhost' to 'localhost5'.
```

**Comment by hesiod** [comment]
I can confirm this bug, running systemd 226 on Linux 4.2.2 and using a systemd-networkd/wpa_supplicant configuration on a laptop (Thinkpad T400).
networkd manages both an ethernet link (enp0s25 in the log) and a WiFi controller (wlp3s0 in the log). The WWAN controller remains unused. The situation occurs when both ethernet and WiFi are connected and routable, so it's probably "just" a DHCP name conflict (both adapters want to claim the same hostname). I followed [the Arch Linux wiki](https://wiki.archlinux.org/index.php/Systemd-networkd#Wired_and_wireless_adapters_on_the_same_machine) and setup RouteMetric for better routing decisions, but that obviously doesn't affect DHCP. The network files set DHCP to yes and enable the IPv6 privacy extensions.

Journal excerpt in a moment.

Just ask for what you need to further nail down the problem.

**Comment by hesiod** [comment]
I removed the IP address and the hostname. Please not that there is no mention of LLMNR responders, as in Keruspe's log.

```
enp0s25: Gained carrier
e1000e: enp0s25 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
enp0s25: DHCPv4 address 192.168.6.101/24 via 192.168.6.1
Detected conflict on some-hostname IN AAAA  [IPv6 address]
Hostname conflict, changing published hostname from 'some-hostname' to 'other-hostname'.
Assertion 'e->key == i->next_key' failed at src/basic/hashmap.c:691, function hashmap_iterate_in_internal_order(). Aborting.
```

**Comment by poettering** [comment]
OK, I am pretty sure this is fixed with f9ebb22. Closing. If you manage to reproduce this with that commit applied, please reopen.
